### ISSUE TITLE
Should ipproto request scope limiting be supported in this draft?

### ISSUE THREAD

**Issue by achernya** [tfpauly/draft-age-masque-connect-ip#12]
Looking at the current text in the draft, we have support for limiting scope to 
* target, which is a hostname or IP address
* IP protocol number

Target clearly makes sense, as that limits the route advertisements and other components that we need to worry about, and allows multiplexing at least the 2-tuple (src, dst) IP addresses.

IP protocol number being supported in this draft is less clear to me. In conjunction with target, it expands our scope to multiplexing 3-tuples (ipproto, src, dst), but that doesn't seem like enough utility over 2-tuple, since the protocols involved are almost always going to be ICMP, TCP, and UDP and I can imagine most users would want to do all of these together.

Given that we're already not supporting TCP/UDP port numbers in this draft, would it make sense to move ipproto out?

**Comment by tfpauly** [comment]
IP protocol is in scope, since it is a field in the IP header, as discussed. That’s the boundary we agreed on.

**Comment by achernya** [comment]
While I agree that the IP protocol is in scope, in scope doesn't mean we have to use it.

My concern is that we can't actually provide the better reuse of addresses on proxies with just protocol number, and we need deeper awareness of how the protocol works. Here's a counter example (and why I think we should defer this feature until we can get richer knowledge):

User connects to example.com, which has an A record for which there is no route. The gateway of the proxy produces an ICMP "Destination unreachable" message. If we have multiplexed the IP to have TCP for user A and ICMP for user B, we would incorrectly deliver the "Destination unreachable" message to the wrong user.

To fix this, the proxy would have to be aware of these protocols at a much deeper level, and introspect over them to be able to multiplex them further. In general because you can get ICMP messages in response to other IP protocol numbers, this multiplexing cannot be done with just protocol number. This is potentially the case for other protocols, but I don't know off the top of my head.

If we look at the ethernet header, it also contains a protocol number field (which is always IP4 or IP6 these days), but all ethernet switching is done using the MAC addresses. The same is true in the IP header for L3 networks, the src addr and dst addr is what's used by all the routers -- the fact that 3 tuple and 5 tuple hashing exists is an optimization to allow for better utilization across multiple links, but in many ways that's an abstraction violation that is taking advantage of an extension point (the protocol number) to do something nice.

**Comment by tfpauly** [comment]
Yes, the ICMP case is one where a proxy would need to be careful.

Stepping back, if we look at the case where you're specifying _both_ a destination and a protocol, it seems pretty clear—saying, I want to speak UDP to this particular host. That shouldn't necessarily stop someone from also speaking TCP to the same host using the same address. I'd prefer to have the specification define how IP protocol is specified, give the caveats about how it can be used (with the ICMP example), and let it be extended.

The line we agreed upon was to have the fields in the IP header be part of the core spec, and I think this clearly fits.

**Comment by achernya** [comment]
We also agreed that the core spec needs to be useful without extensions, and I am unconvinced that ipproto can be used safely without extensions, or without introducing a lot of complexity for ICMP. I think that is in and of itself a good case for considering deferring it.

If we do want to allow different concurrent users to use the same IP for TCP and UDP, we need to handle ICMP specially, even if we do not allow a 3rd user to claim the ICMP ipproto. At that point, we may as well go one step further, and support TCP and UDP's ports --- but we considered that out of scope.

**Comment by DavidSchinazi** [comment]
I'd like to understand what makes ip_proto useful – is anyone planning on building something that wouldn't be possible without ip_proto?

**Comment by mirjak** [comment]
Regarding ICMP I think we anyway need to make sure we have a way to handle icmp correctly. Some of the icmp information are actually relevant for the proxy and therefore forwarding to the client blindly is not always right (except the client asked that explicitly)

**Comment by mirjak** [comment]
I opened a new issue for ICMP handling #21. Further this issue is probably also related to #8.
